src/api/che-login-manager.ts

```diff
@@ -0,0 +1,529 @@
+      // Delete specific login record if any
+      const serverLogins = this.loginData.logins[apiUrl]
+      if (!serverLogins) {
+        // No logins for specified server
+        return
+      }
+      delete serverLogins[username]
+      if (Object.keys(serverLogins).length < 1) {
+        // Delete server without logins
+        delete this.loginData.logins[apiUrl]
+      }
+    }
+
+    if (apiUrl === this.apiUrl) {
+      // Current login info should be deleted
+      this.loginData.lastLoginUrl = this.apiUrl = ''
+      this.loginData.lastUserName = this.username = ''
+    }
+    this.removeExpiredLogins()
+    this.saveLoginData()
+  }
+
+  private readLoginData(): void {
+    if (fs.existsSync(this.dataFilePath)) {
+      this.loginData = JSON.parse(fs.readFileSync(this.dataFilePath).toString()) as CheServerLoginConfig
+    } else {
+      this.loginData = {}
+    }
+
+    if (!this.loginData.logins) {
+      this.loginData.logins = {}
+    }
+  }
+
+  private saveLoginData(): void {
+    this.loginData.lastLoginUrl = this.apiUrl
+    this.loginData.lastUserName = this.username
+    fs.writeFileSync(this.dataFilePath, JSON.stringify(this.loginData))
+  }
+
+  /**
+   * Searches for login data by API URL and user name.
+   * Returns undefined if nothing found by given keys.
+   */
+  private getLoginRecord(apiUrl: string, username: string): RefreshTokenLoginRecord | undefined {
+    const serverLogins = this.loginData.logins![apiUrl]
+    if (!serverLogins) {
+      return
+    }
+    return serverLogins[username]
+  }
+
+  /**
+   * Sets current login credentials by given API URL and username.
+   * If loginRecord is provided, then a new credentials are added, replacing existing if any.
+   * This method doesn't check credentials validity.
+   * Returns true if operation was successful.
+   */
+  private setCurrentLoginContext(apiUrl: string, username: string, loginRecord?: RefreshTokenLoginRecord): boolean {
+    if (!loginRecord) {
+       // Find existing login context and make current
+      loginRecord = this.getLoginRecord(apiUrl, username)
+      if (!loginRecord) {
+        return false
+      }
+    } else {
+      // Set given login config as current
+      let serverLogins = this.loginData.logins![apiUrl]
+      if (!serverLogins) {
+        serverLogins = {}
+        this.loginData.logins![apiUrl] = serverLogins
+      }
+      serverLogins[username] = loginRecord
+    }
+
+    this.apiUrl = apiUrl
+    this.username = username
+    return true
+  }
+
+  private removeExpiredLogins(): void {
+    if (!this.loginData.logins) {
+      return
+    }
+
+    const now = Date.now() / 1000
+    for (const [apiUrl, serverLogins] of Object.entries(this.loginData.logins)) {
+      for (const [username, loginRecord] of Object.entries(serverLogins)) {
+        const decodedToken = jwt.decode(loginRecord.refreshToken)
+        if (decodedToken && typeof decodedToken !== 'string' && decodedToken.exp && decodedToken.exp <= now) {
```

Comment:
I am afraid about the case when server time is different from the local time.
We might immediately remove just obtained login sessions
Is it better just remove `removeExpiredLogins` functionality ?

Reply:
Good point about time zone. However we cannot delete this functionality as we'll end up with a lot of outdated logins!